File: apps/backend/agents/custom_fields_agent.py
"""Custom Fields Agent for extracting structured data from documents."""

from typing import Any

from langchain_core.messages import HumanMessage

from agents.base import get_large_model, get_small_model
from agents.prompts import load_prompt
from config import get_settings
from models.analysis import ConfirmationResult, CustomFieldsAnalysis
from services.paperless import PaperlessClient
from services.qdrant import QdrantService


class CustomFieldsAgent:
    """Agent for extracting custom field values from documents."""

    def __init__(self):
        self.settings = get_settings()
        self.large_model = get_large_model()
        self.small_model = get_small_model()
        self.paperless = PaperlessClient(
            self.settings.paperless_url,
            self.settings.paperless_token,
        )
        # Qdrant is optional - only used for similar document context
        self.qdrant: QdrantService | None = None
        if self.settings.vector_search_enabled:
            self.qdrant = QdrantService(
                qdrant_url=self.settings.qdrant_url,
                collection_name=self.settings.qdrant_collection,
                ollama_url=self.settings.ollama_url,
                embedding_model=self.settings.ollama_embedding_model,
            )

    async def process(self, doc_id: int, content: str) -> dict[str, Any]:
        """Process document to extract custom field values.

        Args:
            doc_id: Document ID
            content: OCR content

        Returns:
            Result dict with custom field values
        """
        # Get custom field definitions
        custom_fields = await self.paperless.get_custom_fields()
        if not custom_fields:
            # No custom fields defined, skip this step
            return {
                "doc_id": doc_id,
                "success": True,
                "skipped": True,
                "reason": "No custom fields defined in Paperless",
            }

        # Get document's current document type for context
        doc = await self.paperless.get_document(doc_id)
        document_type = None
        if doc and doc.get("document_type"):
            document_type = await self.paperless.get_document_type_name(doc["document_type"])

        # Get similar documents (optional)
        similar_docs: list[dict] = []
        if self.qdrant:
            try:
                await self.qdrant.initialize()
                similar_docs = await self.qdrant.search_similar(content[:2000], k=5)
            except Exception:
                # Continue without similar docs - not critical
                pass

        max_retries = self.settings.confirmation_max_retries
        feedback = None

        for attempt in range(max_retries):
            # Analyze custom fields
            analysis = await self._analyze_custom_fields(
                content,
                custom_fields,
                document_type,
                similar_docs,
                feedback,
            )

            # If no fields suggested, we're done
            if not analysis.suggested_fields:
                return {
                    "doc_id": doc_id,
                    "success": True,
                    "fields": [],
                    "reasoning": analysis.reasoning,
                    "attempts": attempt + 1,
                }

            # Confirm with smaller model
            confirmation = await self._confirm_custom_fields(content, analysis)

            if confirmation.confirmed:
                result = await self._apply_custom_fields(doc_id, analysis, custom_fields)
                return {
                    "doc_id": doc_id,
                    "success": True,
                    "fields": [f.model_dump() for f in analysis.suggested_fields],
                    "reasoning": analysis.reasoning,
                    "attempts": attempt + 1,
                    **result,
                }

            feedback = confirmation.feedback

        # Needs review - but for custom fields we don't block the pipeline
        return {
            "doc_id": doc_id,
            "success": True,  # Still success, just not applied
            "needs_review": True,
            "suggested_fields": [f.model_dump() for f in analysis.suggested_fields],
            "reasoning": analysis.reasoning,
            "last_feedback": feedback,
            "attempts": max_retries,
        }

    async def _analyze_custom_fields(
        self,
        content: str,
        custom_fields: list[dict],
        document_type: str | None,
        similar_docs: list[dict],
        feedback: str | None = None,
    ) -> CustomFieldsAnalysis:
        """Analyze document to extract custom field values."""
        prompt_template = load_prompt("custom_fields") or self._default_prompt()

        # Format custom fields with their types
        fields_info = []
        for field in custom_fields:
            field_type = field.get("data_type", "string")
            fields_info.append(f"- ID: {field['id']}, Name: {field['name']}, Type: {field_type}")
        fields_list = "\n".join(fields_info)

        # Format similar docs with their custom fields
        similar_info = []
        for doc in similar_docs[:3]:
            doc_fields = doc.get("metadata", {}).get("custom_fields", [])
            if doc_fields:
                fields_str = ", ".join(f"{f.get('name')}: {f.get('value')}" for f in doc_fields)
                similar_info.append(f"- {doc['metadata'].get('title', 'Unknown')}: {fields_str}")

        # Format the prompt with variables
        formatted_prompt = prompt_template.format(
            document_content=content[:3000],
            document_type=document_type or "Unknown",
            custom_fields=fields_list or "No custom fields defined.",
            similar_docs="\n".join(similar_info) if similar_info else "No similar documents found.",
            feedback=feedback or "None",
        )

        messages = [HumanMessage(content=formatted_prompt)]

        structured_model = self.large_model.with_structured_output(CustomFieldsAnalysis)
        return await structured_model.ainvoke(messages)

    async def _confirm_custom_fields(
        self,
        content: str,
        analysis: CustomFieldsAnalysis,
    ) -> ConfirmationResult:
        """Confirm custom fields with smaller model."""
        confirmation_prompt = load_prompt("confirmation") or self._default_confirmation_prompt()

        # Format analysis result
        fields_summary = "\n".join(
            f"- {f.field_name}: {f.value} (Reason: {f.reasoning})"
            for f in analysis.suggested_fields
        )

        analysis_result = f"""**Suggested Custom Fields:**
{fields_summary if fields_summary else "No fields suggested"}

**Overall Reasoning:** {analysis.reasoning}"""

        # Format the prompt with variables
        formatted_prompt = confirmation_prompt.format(
            analysis_result=analysis_result,
            document_excerpt=content[:1500],
        )

        messages = [HumanMessage(content=formatted_prompt)]

        structured_model = self.small_model.with_structured_output(ConfirmationResult)
        return await structured_model.ainvoke(messages)

    async def _apply_custom_fields(
        self,
        doc_id: int,
        analysis: CustomFieldsAnalysis,
        custom_fields: list[dict],
    ) -> dict[str, Any]:
        """Apply custom field values to document."""
        # Get current document custom fields
        doc = await self.paperless.get_document(doc_id)
        current_fields = doc.get("custom_fields", []) if doc else []

        # Build field ID mapping
        field_id_to_def = {f["id"]: f for f in custom_fields}

        # Update custom fields
        updated_fields = []
        for field_value in analysis.suggested_fields:
            field_def = field_id_to_def.get(field_value.field_id)
            if not field_def:
                continue

            # Find or create field entry
            found = False
            for cf in current_fields:
                if cf.get("field") == field_value.field_id:
                    cf["value"] = field_value.value
                    found = True
                    break

            if not found:
                current_fields.append(
                    {
                        "field": field_value.field_id,
                        "value": field_value.value,
                    }
                )

            updated_fields.append(field_value.field_name)

        # Update document
        if updated_fields:
            await self.paperless.update_document(doc_id, custom_fields=current_fields)

        return {
            "applied": bool(updated_fields),
            "updated_fields": updated_fields,
        }

    def _default_prompt(self) -> str:
        return """You are a document data extraction specialist. Extract structured information into custom fields.

Document Type: {document_type}

## Available Custom Fields

{custom_fields}

## Document Content

{document_content}

## Similar Documents

{similar_docs}

## Previous Feedback

{feedback}

Extract any values that match the available custom fields. Only extract values that are clearly present in the document."""

    def _default_confirmation_prompt(self) -> str:
        return """You are a quality assurance assistant. Review the custom field extraction.

Consider:
- Are the values actually present in the document?
- Are the values correctly formatted for the field type?
- Are there any extraction errors?

Accept if the values look reasonable. Reject only if there are clear errors."""

File: apps/backend/src/agents/CustomFieldsAgent.ts
/**
 * Custom Fields extraction agent.
 */
import { Effect, Context, Layer, Stream } from 'effect';
import { ConfigService, OllamaService, PromptService, TinyBaseService, PaperlessService } from '../services/index.js';
import { AgentError } from '../errors/index.js';
import type { CustomField, CustomFieldValue } from '../models/index.js';
import {
  type Agent,
  type AgentProcessResult,
  type StreamEvent,
  runConfirmationLoop,
  emitStart,
  emitThinking,
  emitAnalyzing,
  emitConfirming,
  emitResult,
  emitComplete,
} from './base.js';

// ===========================================================================
// Types
// ===========================================================================

export interface CustomFieldsInput {
  docId: number;
  content: string;
  documentType?: string;
  customFields: CustomField[];
}

export interface FieldValue {
  fieldId: number;
  fieldName: string;
  value: string | number | boolean | null;
  reasoning: string;
}

export interface CustomFieldsAnalysis {
  suggestedFields: FieldValue[];
  reasoning: string;
  confidence: number;
}

export interface CustomFieldsResult extends AgentProcessResult {
  fields: FieldValue[];
  updatedFields: string[];
  skipped?: boolean;
  skipReason?: string;
}

// ===========================================================================
// Service Interface
// ===========================================================================

export interface CustomFieldsAgentService extends Agent<CustomFieldsInput, CustomFieldsResult> {
  readonly name: 'custom_fields';
  readonly process: (input: CustomFieldsInput) => Effect.Effect<CustomFieldsResult, AgentError>;
  readonly processStream: (input: CustomFieldsInput) => Stream.Stream<StreamEvent, AgentError>;
}

export const CustomFieldsAgentService = Context.GenericTag<CustomFieldsAgentService>('CustomFieldsAgentService');

// ===========================================================================
// Response Parsers
// ===========================================================================

const parseAnalysisResponse = (response: string, customFields: CustomField[]): CustomFieldsAnalysis => {
  const fieldIdToName = new Map(customFields.map((f) => [f.id, f.name]));

  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      const parsed = JSON.parse(jsonMatch[0]) as {
        suggested_fields?: Array<{
          field_id?: number;
          fieldId?: number;
          field_name?: string;
          fieldName?: string;
          value?: string | number | boolean | null;
          reasoning?: string;
        }>;
        suggestedFields?: Array<{
          field_id?: number;
          fieldId?: number;
          field_name?: string;
          fieldName?: string;
          value?: string | number | boolean | null;
          reasoning?: string;
        }>;
        reasoning?: string;
        confidence?: number;
      };

      const fields = parsed.suggested_fields ?? parsed.suggestedFields ?? [];
      const suggestedFields = fields.map((f) => {
        const fieldId = f.field_id ?? f.fieldId ?? 0;
        return {
          fieldId,
          fieldName: f.field_name ?? f.fieldName ?? fieldIdToName.get(fieldId) ?? '',
          value: f.value ?? null,
          reasoning: f.reasoning ?? '',
        };
      });

      return {
        suggestedFields,
        reasoning: parsed.reasoning ?? '',
        confidence: parsed.confidence ?? 0.5,
      };
    }
  } catch {
    // Fall back to empty
  }

  return {
    suggestedFields: [],
    reasoning: 'Could not parse response',
    confidence: 0,
  };
};

const parseConfirmationResponse = (response: string): { confirmed: boolean; feedback?: string } => {
  const lower = response.toLowerCase();
  const confirmed = lower.includes('yes') || lower.includes('confirmed') || lower.includes('accept');
  const feedback = confirmed ? undefined : response.replace(/^(no|rejected|feedback:)\s*/i, '');
  return { confirmed, feedback };
};

// ===========================================================================
// Live Implementation
// ===========================================================================

export const CustomFieldsAgentServiceLive = Layer.effect(
  CustomFieldsAgentService,
  Effect.gen(function* () {
    const config = yield* ConfigService;
    const ollama = yield* OllamaService;
    const prompts = yield* PromptService;
    const tinybase = yield* TinyBaseService;
    const paperless = yield* PaperlessService;

    const { autoProcessing, tags: tagConfig } = config.config;

    return {
      name: 'custom_fields' as const,

      process: (input: CustomFieldsInput) =>
        Effect.gen(function* () {
          // If no custom fields defined, skip
          if (input.customFields.length === 0) {
            return {
              success: true,
              value: null,
              reasoning: 'No custom fields defined',
              confidence: 1,
              alternatives: [],
              attempts: 0,
              needsReview: false,
              fields: [],
              updatedFields: [],
              skipped: true,
              skipReason: 'No custom fields defined in Paperless',
            };
          }

          return yield* runConfirmationLoop<CustomFieldsAnalysis, CustomFieldsResult>({
            maxRetries: autoProcessing.confirmationMaxRetries,

            analyze: (feedback) =>
              Effect.gen(function* () {
                // Format custom fields for prompt
                const fieldsInfo = input.customFields
                  .map((f) => `- ID: ${f.id}, Name: ${f.name}, Type: ${f.data_type}`)
                  .join('\n');

                const prompt = yield* prompts.renderPrompt('custom_fields', {
                  document_content: input.content.slice(0, 8000),
                  document_type: input.documentType ?? 'Unknown',
                  custom_fields: fieldsInfo,
                  feedback: feedback ?? 'None',
                });

                const response = yield* ollama.generate(
                  ollama.getModel('large'),
                  prompt,
                  { temperature: 0.1 }
                );

                return parseAnalysisResponse(response, input.customFields);
              }).pipe(
                Effect.mapError((e) =>
                  new AgentError({
                    message: `Custom fields analysis failed: ${e}`,
                    agent: 'custom_fields',
                    step: 'analyze',
                    cause: e,
                  })
                )
              ),

            confirm: (analysis) =>
              Effect.gen(function* () {
                if (analysis.suggestedFields.length === 0) {
                  return { confirmed: true }; // No fields to extract is valid
                }

                const fieldsSummary = analysis.suggestedFields
                  .map((f) => `${f.fieldName}: ${f.value} (Reason: ${f.reasoning})`)
                  .join('\n');

                const prompt = yield* prompts.renderPrompt('custom_fields_confirmation', {
                  document_excerpt: input.content.slice(0, 4000),
                  suggested_fields: fieldsSummary,
                  reasoning: analysis.reasoning,
                });

                const response = yield* ollama.generate(
                  ollama.getModel('small'),
                  prompt,
                  { temperature: 0 }
                );

                return parseConfirmationResponse(response);
              }).pipe(
                Effect.mapError((e) =>
                  new AgentError({
                    message: `Custom fields confirmation failed: ${e}`,
                    agent: 'custom_fields',
                    step: 'confirm',
                    cause: e,
                  })
                )
              ),

            apply: (analysis) =>
              Effect.gen(function* () {
                if (analysis.suggestedFields.length === 0) {
                  return {
                    success: true,
                    value: null,
                    reasoning: analysis.reasoning,
                    confidence: analysis.confidence,
                    alternatives: [],
                    attempts: 1,
                    needsReview: false,
                    fields: [],
                    updatedFields: [],
                  };
                }

                // Get current document
                const doc = yield* paperless.getDocument(input.docId);
                const currentFields = (doc.custom_fields ?? []) as CustomFieldValue[];

                const fieldIdToField = new Map(input.customFields.map((f) => [f.id, f]));
                const updatedFields: string[] = [];

                // Build new custom fields array
                const newCustomFields: CustomFieldValue[] = [...currentFields];

                for (const fieldValue of analysis.suggestedFields) {
                  const fieldDef = fieldIdToField.get(fieldValue.fieldId);
                  if (!fieldDef) continue;

                  // Find or create field entry
                  const existingIdx = newCustomFields.findIndex(
                    (cf) => cf.field === fieldValue.fieldId
                  );

                  if (existingIdx >= 0) {
                    newCustomFields[existingIdx] = {
                      field: fieldValue.fieldId,
                      value: fieldValue.value,
                    };
                  } else {
                    newCustomFields.push({
                      field: fieldValue.fieldId,
                      value: fieldValue.value,
                    });
                  }

                  updatedFields.push(fieldValue.fieldName);
                }

                // Update document
                if (updatedFields.length > 0) {
                  yield* paperless.updateDocument(input.docId, {
                    custom_fields: newCustomFields,
                  });
                }

                return {
                  success: true,
                  value: updatedFields.join(', '),
                  reasoning: analysis.reasoning,
                  confidence: analysis.confidence,
                  alternatives: [],
                  attempts: 1,
                  needsReview: false,
                  fields: analysis.suggestedFields,
                  updatedFields,
                };
              }).pipe(
                Effect.mapError((e) =>
                  new AgentError({
                    message: `Custom fields application failed: ${e}`,
                    agent: 'custom_fields',
                    step: 'apply',
                    cause: e,
                  })
                )
              ),

            onMaxRetries: (lastAnalysis) =>
              Effect.gen(function* () {
                // Custom fields don't block the pipeline
                return {
                  success: true,
                  value: null,
                  reasoning: lastAnalysis.reasoning,
                  confidence: lastAnalysis.confidence,
                  alternatives: [],
                  attempts: autoProcessing.confirmationMaxRetries,
                  needsReview: true,
                  fields: lastAnalysis.suggestedFields,
                  updatedFields: [],
                };
              }).pipe(
                Effect.mapError((e) =>
                  new AgentError({
                    message: `Custom fields max retries: ${e}`,
                    agent: 'custom_fields',
                    step: 'max_retries',
                    cause: e,
                  })
                )
              ),
          });
        }).pipe(
          Effect.mapError((e) =>
            e instanceof AgentError
              ? e
              : new AgentError({
                  message: `Custom fields processing failed: ${e}`,
                  agent: 'custom_fields',
                  cause: e,
                })
          )
        ),

      processStream: (input: CustomFieldsInput) =>
        Stream.asyncEffect<StreamEvent, AgentError>((emit) =>
          Effect.gen(function* () {
            yield* Effect.sync(() => emit.single(emitStart('custom_fields')));

            // If no custom fields defined, skip
            if (input.customFields.length === 0) {
              yield* Effect.sync(() =>
                emit.single(
                  emitResult('custom_fields', {
                    success: true,
                    skipped: true,
                    reason: 'No custom fields defined',
                  })
                )
              );
              yield* Effect.sync(() => emit.single(emitComplete('custom_fields')));
              yield* Effect.sync(() => emit.end());
              return;
            }

            let feedback: string | null = null;
            let lastAnalysis: CustomFieldsAnalysis | null = null;

            for (let attempt = 0; attempt < autoProcessing.confirmationMaxRetries; attempt++) {
              yield* Effect.sync(() =>
                emit.single(emitAnalyzing('custom_fields', `Attempt ${attempt + 1}`))
              );

              const fieldsInfo = input.customFields
                .map((f) => `- ID: ${f.id}, Name: ${f.name}, Type: ${f.data_type}`)
                .join('\n');

              const prompt = yield* prompts.renderPrompt('custom_fields', {
                document_content: input.content.slice(0, 8000),
                document_type: input.documentType ?? 'Unknown',
                custom_fields: fieldsInfo,
                feedback: feedback ?? 'None',
              });

              const analysisResponse = yield* ollama.generate(
                ollama.getModel('large'),
                prompt,
                { temperature: 0.1 }
              );

              const analysis = parseAnalysisResponse(analysisResponse, input.customFields);
              lastAnalysis = analysis;

              yield* Effect.sync(() =>
                emit.single(emitThinking('custom_fields', analysis.reasoning))
              );

              if (analysis.suggestedFields.length === 0) {
                yield* Effect.sync(() =>
                  emit.single(
                    emitResult('custom_fields', {
                      success: true,
                      fields: [],
                      updatedFields: [],
                    })
                  )
                );
                yield* Effect.sync(() => emit.single(emitComplete('custom_fields')));
                yield* Effect.sync(() => emit.end());
                return;
              }

              const fieldsSummary = analysis.suggestedFields
                .map((f) => `${f.fieldName}: ${f.value}`)
                .join(', ');

              yield* Effect.sync(() =>
                emit.single(emitConfirming('custom_fields', fieldsSummary))
              );

              const confirmPrompt = yield* prompts.renderPrompt('custom_fields_confirmation', {
                document_excerpt: input.content.slice(0, 4000),
                suggested_fields: fieldsSummary,
                reasoning: analysis.reasoning,
              });

              const confirmResponse = yield* ollama.generate(
                ollama.getModel('small'),
                confirmPrompt,
                { temperature: 0 }
              );

              const confirmation = parseConfirmationResponse(confirmResponse);

              if (confirmation.confirmed) {
                const doc = yield* paperless.getDocument(input.docId);
                const currentFields = (doc.custom_fields ?? []) as CustomFieldValue[];
                const fieldIdToField = new Map(input.customFields.map((f) => [f.id, f]));
                const updatedFields: string[] = [];

                const newCustomFields: CustomFieldValue[] = [...currentFields];

                for (const fieldValue of analysis.suggestedFields) {
                  const fieldDef = fieldIdToField.get(fieldValue.fieldId);
                  if (!fieldDef) continue;

                  const existingIdx = newCustomFields.findIndex(
                    (cf) => cf.field === fieldValue.fieldId
                  );

                  if (existingIdx >= 0) {
                    newCustomFields[existingIdx] = {
                      field: fieldValue.fieldId,
                      value: fieldValue.value,
                    };
                  } else {
                    newCustomFields.push({
                      field: fieldValue.fieldId,
                      value: fieldValue.value,
                    });
                  }

                  updatedFields.push(fieldValue.fieldName);
                }

                if (updatedFields.length > 0) {
                  yield* paperless.updateDocument(input.docId, {
                    custom_fields: newCustomFields,
                  });
                }

                yield* Effect.sync(() =>
                  emit.single(
                    emitResult('custom_fields', {
                      success: true,
                      fields: analysis.suggestedFields,
                      updatedFields,
                    })
                  )
                );

                yield* Effect.sync(() => emit.single(emitComplete('custom_fields')));
                yield* Effect.sync(() => emit.end());
                return;
              }

              feedback = confirmation.feedback ?? 'Not confirmed';
            }

            // Max retries - custom fields don't block pipeline
            yield* Effect.sync(() =>
              emit.single(
                emitResult('custom_fields', {
                  success: true,
                  needsReview: true,
                  fields: lastAnalysis?.suggestedFields ?? [],
                  updatedFields: [],
                })
              )
            );

            yield* Effect.sync(() => emit.single(emitComplete('custom_fields')));
            yield* Effect.sync(() => emit.end());
          }).pipe(
            Effect.mapError((e) =>
              new AgentError({
                message: `Custom fields stream failed: ${e}`,
                agent: 'custom_fields',
                cause: e,
              })
            )
          )
        ),
    };
  })
);
